design/one-pager-crd-scaling.md

```diff
@@ -0,0 +1,563 @@
+alone. [Azure Service Operator][aso], or [GCP Config
+Connector][config-connector] are projects that rely on Kubernetes [custom
+resources] as an extension mechanism and have many CRDs representing associated
+Cloud resources. 
+
+Kubernetes is a complex ecosystem with many moving parts and we need a deeper
+understanding for the issues around scaling in the dimension of the total number
+of CRDs per-cluster. This dimension is not yet officially considered in the
+scalability thresholds document [[3]] but it will be with good probability. So,
+as the Crossplane community, we would like to have our use cases considered in
+relevant contexts, and we would like to gain a good understanding so that we
+can:
+- Establish a common understanding around relevant components and issues
+- Discuss and contribute to relevant upstream discussions with the broader
+  community
+- Come up with tooling that can help reproducing the issues and assessing the
+  effectiveness of considered solutions
+- Establish a common understanding on the Crossplane scenarios and if possible,
+  a set of “definition of done” criterias for the CRD scaling issues, i.e., have
+  a clear expectation on the scenarios we would like to support and on the
+  expected performance in those scenarios.
+
+ 
+
+## Issues
+
+We can categorize the issues that we observe when scaling a cluster in the
+number of installed CRDs in two:
+- Client-side issues: Clients of the control-plane (`kubectl`, controllers,
+  etc.) experience extended delays in the requests they make. So far, our
+  observation is that all the issues that we categorize as client-side issues
+  are due to “client-side throttling”, and not caused by the API server itself.
+  These clients can throttle the requests they make to the API server solely as
+  a function of the [rate limiter] implementation they are using with no
+  feedback from the server. Per our observations, the problematic client-side
+  component that’s responsible for the high number of requests in a unit time
+  (high request rate) is the [discovery client]. In the [Client-side
+  Throttling](#client-side-throttling) section below, we will examine how
+  `kubectl` behaves when there is a large number of CRDs installed in the
+  cluster and how the discovery cache and the discovery client affect the
+  perceived performance of the `kubectl` commands run. 
+- Server-side issues: We also have some observations on the control-plane
+  components such as increased resource consumption in `kube-apiserver`
+  proportional to the number of CRDs we install in the cluster. While we expect
+  more resources to be allocated as we install more CRDs, some detailed analysis
+  presented in the [API Server Resource
+  Consumption](#api-server-resource-consumption) section reveals some possible
+  improvement points and some other points such as OpenAPI v3 schema sizes where
+  we should be careful.
+
+
+### Client-side Throttling
```

Comment:
is it possible to very clearly state (maybe with a table?) which versions of kubectl are released (or targeting releases) that have all the client-side throttling fixes and we can recommend to the community?

Reply:
e.g. it's not super obvious to me (as a casual reader) which versions of kubectl (and also k8s-api) i would need to have a good experience with jet providers.  Maybe we can make that super obvious at the top of the doc so folks coming to this document just looking for guidance on what versions to use to have a good experience can easily find that info without getting lost in the details.  what do you think @ulucinar?

Reply:
I think it's a good idea @jbw976. Added a `kubectl` version table with their release dates as a recommendation at the beginning of the `Client-side Throttling` section.

Reply:
Thanks @jbw976. I've also added a table of kube-apiserver versions to the beginning of the `API Server Resource Consumption` section explicitly mentioning about the OpenAPI v2 spec lazy-marshaling change and the related upstream issues.

Reply:
that is super helpful @ulucinar, thank you for making this document more accessible for a broader audience!! 🙇‍♂️